libmamba/src/solv-cpp/repo.cpp

```diff
@@ -0,0 +1,278 @@
+
+namespace mamba::solv
+{
+    /****************************************
+     *  Implementation of ConstObjRepoView  *
+     ****************************************/
+
+    ObjRepoViewConst::ObjRepoViewConst(const ::Repo* repo) noexcept
+        : m_repo{ repo }
+    {
+    }
+
+    auto ObjRepoViewConst::raw() const -> const ::Repo*
+    {
+        return m_repo;
+    }
+
+    auto ObjRepoViewConst::id() const -> RepoId
+    {
+        // No function exists to get a repo id
+        const ::Pool* pool = raw()->pool;
+        const ::Repo* repo = nullptr;
+        int repoid = 0;
+        FOR_REPOS(repoid, repo)
+        {
+            if (repo == raw())
+            {
+                return repoid;
+            }
+        }
+        assert(false);
+        return 0;
+    }
+
+    auto ObjRepoViewConst::n_solvables() const -> std::size_t
+    {
+        assert(raw()->nsolvables >= 0);
+        return static_cast<std::size_t>(raw()->nsolvables);
+    }
+
+    namespace
+    {
+        auto get_solvable_ptr(const ::Repo* repo, SolvableId id) -> ::Solvable*
+        {
+            if ((id >= repo->start) && (id < repo->end))
+            {
+                if (Solvable* const s = ::pool_id2solvable(repo->pool, id); s != nullptr)
+                {
+                    if (s->repo == repo)
+                    {
+                        return s;
+                    }
+                }
+            }
+            return nullptr;
+        }
+    }
+
+    auto ObjRepoViewConst::has_solvable(SolvableId id) const -> bool
+    {
+        return get_solvable_ptr(raw(), id) != nullptr;
+    }
+
+    auto ObjRepoViewConst::get_solvable(SolvableId id) const -> std::optional<ObjSolvableViewConst>
+    {
+        if (const ::Solvable* s = get_solvable_ptr(raw(), id); s != nullptr)
+        {
+            return { ObjSolvableViewConst{ s } };
+        }
+        return std::nullopt;
+    }
+
+    void ObjRepoViewConst::write(std::filesystem::path solv_file) const
+    {
+        auto* fp = std::fopen(solv_file.string().c_str(), "wb");
+        if (!fp)
+        {
+            throw std::system_error(errno, std::generic_category());
+        }
+        const auto write_res = ::repo_write(const_cast<::Repo*>(raw()), fp);
+        const auto close_res = std::fclose(fp);
+        if (write_res != 0)
+        {
+            // TODO(C++20) fmt::format
+            auto ss = std::stringstream();
+            ss << "Unable to write repo '" << name() << "' to file";
+            throw std::runtime_error(ss.str());
+        }
+        if (close_res != 0)
+        {
+            // TODO(C++20) fmt::format
+            auto ss = std::stringstream();
+            ss << "Unable to write file '" << solv_file << '\'';
+            throw std::runtime_error(ss.str());
+        }
+    }
+
+    /***********************************
+     *  Implementation of ObjRepoView  *
+     ***********************************/
+
+    ObjRepoView::ObjRepoView(::Repo* repo) noexcept
```

Comment:
BTW is `nullptr` valid here?

Reply:
Hmm, I'd say not. What should it do, throw? Same remark for Solvables.